Andrew Pinski

ChangeLog:

  * dwarf2out.c (dwarf2out_switch_text_section): Export.
  * debug.h (dwarf2out_switch_text_section): Prototype.
  * final.c (final_scan_insn): Call dwarf2out_switch_text_section
    if we need to call it for dwarf2 unwind info.

diff -ur gcc-4.1-20051105-orig/gcc/debug.h gcc-4.1-20051105/gcc/debug.h
--- gcc-4.1-20051105-orig/gcc/debug.h	2005-11-12 00:09:33.376787809 +0000
+++ gcc-4.1-20051105/gcc/debug.h	2005-11-12 00:10:06.962349214 +0000
@@ -160,6 +160,7 @@
 /* Decide whether we want to emit frame unwind information for the current
    translation unit.  */
 extern int dwarf2out_do_frame (void);
+void dwarf2out_switch_text_section (void);
 
 extern void debug_flush_symbol_queue (void);
 extern void debug_queue_symbol (tree);
diff -ur gcc-4.1-20051105-orig/gcc/dwarf2out.c gcc-4.1-20051105/gcc/dwarf2out.c
--- gcc-4.1-20051105-orig/gcc/dwarf2out.c	2005-11-12 00:09:26.948488952 +0000
+++ gcc-4.1-20051105/gcc/dwarf2out.c	2005-11-12 00:10:06.962349214 +0000
@@ -3520,7 +3520,6 @@
 static void dwarf2out_abstract_function (tree);
 static void dwarf2out_var_location (rtx);
 static void dwarf2out_begin_function (tree);
-static void dwarf2out_switch_text_section (void);
 
 /* The debug hooks structure.  */
 
@@ -6866,7 +6865,7 @@
   *d = new_loc_list (descr, begin, end, section, 0);
 }
 
-static void
+void
 dwarf2out_switch_text_section (void)
 {
   dw_fde_ref fde;
diff -ur gcc-4.1-20051105-orig/gcc/final.c gcc-4.1-20051105/gcc/final.c
--- gcc-4.1-20051105-orig/gcc/final.c	2005-11-12 00:09:26.912487279 +0000
+++ gcc-4.1-20051105/gcc/final.c	2005-11-12 00:10:06.962349214 +0000
@@ -1717,11 +1717,21 @@
 	  if (last_text_section == in_text)
 	    {
 	      (*debug_hooks->switch_text_section) ();
+#if defined (DWARF2_UNWIND_INFO)
+	        if (write_symbols != DWARF2_DEBUG && write_symbols != VMS_AND_DWARF2_DEBUG
+	 	    && dwarf2out_do_frame ())
+	 	  dwarf2out_switch_text_section ();
+#endif
 	      unlikely_text_section ();
 	    }
 	  else
 	    {
 	      (*debug_hooks->switch_text_section) ();
+#if defined (DWARF2_UNWIND_INFO)
+	        if (write_symbols != DWARF2_DEBUG && write_symbols != VMS_AND_DWARF2_DEBUG
+	 	    && dwarf2out_do_frame ())
+	 	  dwarf2out_switch_text_section ();
+#endif
 	      text_section ();
 	    }
 	  break;
